### ISSUE TITLE
False positive InvalidArgument for preg_replace_callback and preg_match

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#2126]
Possibly related to changes to Psalm's handling of functions with alternates

```php
<?php
function f(array &$configs, string $value) {
    // NOTE: Adding a return type would fix the InvalidArgument error
    // Observed: InvalidArgument - Argument 3 of preg_replace_callback expects array<array-key, string>, string provided
    // Expected: Should take closest match for callable and not emit InvalidArgument
    $value = preg_replace_callback('/{{{(\w*)}}}/', function (array $match) use ($configs) {
        $key = $match[1];
        return $configs[$key];
    }, $value, -1, $count);
}

```

```
<?php
/** @param int[]|string[]|bool[] $params */
function test($params) {
    // Related:
    // ERROR: InvalidScalarArgument - src/match.php:4:77 - Argument 2 of preg_match expects string, int|string|bool provided
    // ERROR: InvalidArgument - src/match.php:4:108 - Argument 4 of preg_match expects int(0), int(256) provided
    // Expected: Should warn about InvalidScalarArgument but not InvalidArgument. All 3 alternate signatures have 'subject' => 'string'
    // Observed: Warns about both
    if (!array_key_exists('x', $params) || preg_match('/\/ab\d+\//', $params['x'], $matches, PREG_OFFSET_CAPTURE) == 0 || !empty($params['is'])) {
    }
}
```

**Comment by muglug** [comment]
It's that last argument that messes things up:

```php
function f(string $value) : string {
    // InvalidArgument
    return preg_replace_callback('/a/', function () { return $_GET["a"]; }, $value, -1, $count);
}
function g(string $value) : string {
    // ok
    return preg_replace_callback('/a/', function () { return $_GET["a"]; }, $value, -1);
}
```

https://psalm.dev/r/1466b6d7e1

**Comment by muglug** [comment]
This is fixed
